independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverTransformer.java

```diff
@@ -0,0 +1,121 @@
+package io.quarkus.arc.processor;
+
+import java.util.Collection;
+import java.util.Set;
+import javax.enterprise.event.Reception;
+import javax.enterprise.event.TransactionPhase;
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.MethodInfo;
+import org.jboss.jandex.Type;
+
+/**
+ * Allows a build-time extension to:
+ * <ul>
+ * <li>override the set of qualifiers and priority of any observer method,</li>
+ * <li>instruct the container to ignore the observer method.</li>
+ * </ul>
+ */
+public interface ObserverTransformer extends BuildExtension {
+
+    /**
+     *
+     * @param observedType
+     * @param qualifiers
+     * @return {@code true} if this transformer is meant to be applied to the supplied attributes of an observer method,
+     *         {@code false} otherwise
+     */
+    boolean appliesTo(Type observedType, Set<AnnotationInstance> qualifiers);
+
+    /**
+     * @param context
+     */
+    void transform(TransformationContext context);
+
+    /**
+     * Transformation context.
+     */
+    interface TransformationContext extends BuildExtension.BuildContext {
+
+        /**
+         * 
+         * @return the corresponding observer method
+         */
+        MethodInfo getMethod();
+
+        /**
+         * 
+         * @return the observed type
+         */
+        Type getObservedType();
+
+        /**
+         *
+         * @return the set of qualifiers
+         */
+        Set<AnnotationInstance> getQualifiers();
+
+        /**
+         * 
+         * @return the reception
+         */
+        Reception getReception();
+
+        /**
+         * 
+         * @return the transaction phase
+         */
+        TransactionPhase getTransactionPhase();
+
+        /**
+         * Retrieves all annotations declared on the observer method. This method is preferred to manual inspection
+         * of {@link #getMethod()} which may, in some corner cases, hold outdated information.
+         * <p>
+         * The resulting set of annotations contains contains annotations that belong to the method itself
+         * as well as to its parameters.
+         *
+         * @return collection of all annotations
+         */
+        Collection<AnnotationInstance> getAllAnnotations();
+
+        /**
+         * Instruct the container to ignore the observer method.
+         * 
+         * @return self
+         */
+        void veto();
+
+        /**
+         * The transformation is not applied until the {@link AnnotationsTransformation#done()} method is invoked.
+         *
+         * @return a new transformation
+         */
+        ObserverTransformation transform();
+
+    }
+
+    interface ObserverTransformation extends AnnotationsTransformation<ObserverTransformation> {
```

Comment:
Hm, I'm not so sure. Async observers are invoked in a different context. But the truth is that it's possible to change the transaction phase which could eventually (transactional observers are not implemented yet) result in an async invocation as well.

Reply:
> Async observers are invoked in a different context. 

What do you mean by different context?

I just thought that since we allow to change reception and tx phases, this could work as well.
It shouldn't change/break the logic inside the observer either.

Reply:
> What do you mean by different context?

I mean different thread, transaction context, CDI contexts...

> we allow to change reception and tx phases

Reception is ok - does not change the context. Transaction phase - yep, that's what I mentioned in the previous comment.